src/docfx/build/redirection/RedirectionProvider.cs

```diff
@@ -167,26 +169,44 @@ private static IEnumerable<string> ProbeRedirectionFiles(string docsetPath)
         private IReadOnlyDictionary<FilePath, FilePath> GetRenameHistory(
             RedirectionItem[] redirections, IReadOnlyDictionary<FilePath, string> redirectUrls)
         {
-            // Convert the redirection target from redirect url to file path according to the version of redirect source
-            var renameHistory = new Dictionary<FilePath, FilePath>();
+            var files = new ConcurrentBag<(FilePath path, string siteUrl, IReadOnlyList<string> monikers)>();
+
+            ParallelUtility.ForEach(_errorLog, _buildScope.Files.Concat(redirectUrls.Keys), file =>
+            {
+                var (error, monikers) = _monikerProvider.GetFileLevelMonikers(file);
+                if (error != null)
+                {
+                    _errorLog.Write(error);
+                }
 
-            var publishUrlMap = _buildScope.Files.Concat(redirectUrls.Keys)
-                .GroupBy(file => _documentProvider.GetDocument(file).SiteUrl)
+                files.Add((file, _documentProvider.GetDocument(file).SiteUrl, monikers));
+            });
+
+            var publishUrlMap = files
+                .GroupBy(file => file.siteUrl)
                 .ToDictionary(group => group.Key, group => group.ToList(), PathUtility.PathComparer);
 
-            foreach (var item in redirections.Where(item => item.RedirectDocumentId))
+            var renameHistory = new ConcurrentDictionary<FilePath, FilePath>();
+
+            // Convert the redirection target from redirect url to file path according to the version of redirect source
+            Parallel.ForEach(redirections, item =>
```

Comment:
>Parallel [](start = 12, length = 8)

How do we decide whether using `Parallel` or `ParallelUtility`?

Reply:
- The action can throw `DocfxException`
- This function should continue in case action throws `DocfxException`

Reply:
`GetFileLevelMonikers` is also invoked in this parallel scope, though logically it wouldn't throw. That's why I was confused.